packages/nextcloud/lib/src/webdav/path_uri.dart

```diff
@@ -0,0 +1,76 @@
+import 'package:collection/collection.dart';
+import 'package:meta/meta.dart';
+
+/// A [Uri] that only exposes the path handling.
+@immutable
+class PathUri {
+  /// Creates a new path URI.
+  ///
+  /// Only one of [path] and [pathSegments] must be given.
+  PathUri({final String? path, final List<String>? pathSegments})
+      : assert(
+          (path == null) != (pathSegments == null) || (path == null && pathSegments == null),
+          'Only one of path and pathSegments must be given',
+        ),
+        _uri = Uri(
+          path: path,
+          pathSegments: pathSegments,
+        ).normalizePath();
+
+  final Uri _uri;
+
+  /// Returns the path as a string.
+  ///
+  /// See [pathSegments] for getting the path as a list of its segments.
+  String get path => _uri.path;
+
+  /// Returns the path as a list of its segments.
+  ///
+  /// See [path] for getting the path as a string.
+  List<String> get pathSegments => _uri.pathSegments;
+
+  /// Returns whether the path is a directory.
+  ///
+  /// If the whole path is empty it is considered to be a directory.
+  bool get isDirectory => pathSegments.lastOrNull?.isEmpty ?? true;
+
+  /// Returns the basename of the path.
+  String get basename => pathSegments.where((final s) => s.isNotEmpty).lastOrNull ?? '';
+
+  /// Returns the dirname of the path.
+  PathUri get dirname {
```

Comment:
The name if this getter is confusing.
I'd expect a dirname to be s String and not a Uri.

I suggest to rename the getter `parentDirectory` and add a separate getter `direname` that calls `parentDirectory.basename`

Reply:
This was based off the `basename` and `dirname` commands, but I agree that `parent` would make more sense.